apps/provider.c

```diff
@@ -0,0 +1,279 @@
+            if (verbose < i)
+                verbose = i;
+            break;
+        }
+    }
+
+    /* Allow any trailing parameters as provider names. */
+    argc = opt_num_rest();
+    argv = opt_rest();
+    for ( ; *argv; argv++) {
+        if (**argv == '-') {
+            BIO_printf(bio_err, "%s: Cannot mix flags and provider names.\n",
+                       prog);
+            BIO_printf(bio_err, "%s: Use -help for summary.\n", prog);
+            goto end;
+        }
+        sk_OPENSSL_CSTRING_push(providers, *argv);
+    }
+
+    ret = 0;
+    for (i = 0; i < sk_OPENSSL_CSTRING_num(providers); i++) {
+        const char *name = sk_OPENSSL_CSTRING_value(providers, i);
+        OSSL_PROVIDER *prov = OSSL_PROVIDER_load(NULL, name);
+
+        if (prov != NULL) {
+            BIO_printf(bio_out, verbose == 0 ? "%s\n" :  "[ %s ]\n", name);
+
+            if (verbose > 0) {
+                META data;
+
+                data.total = 0;
+                data.first = 1;
+                data.verbose = verbose;
+                data.prov = prov;
+                data.fn = print_caps;
+
+                switch (verbose) {
+                case 1:
+                    BIO_printf(bio_out, "    ");
+                    break;
+                case 2:
+                    data.indent = 4;
+                    break;
+                case 3:
+                default:
+                    data.indent = 4;
+                    data.subindent = 10;
+                    break;
+                }
+
+                if (verbose > 1) {
+                    data.first = 1;
+                    data.lable = "Cipher";
+                }
+                EVP_CIPHER_do_all_ex(NULL, do_cipher, &data);
+                if (verbose > 1) {
+                    data.first = 1;
+                    data.lable = "Digest";
+                }
+                EVP_MD_do_all_ex(NULL, do_digest, &data);
+                if (verbose > 1) {
+                    data.first = 1;
+                    data.lable = "MAC";
+                }
+                EVP_MAC_do_all_ex(NULL, do_mac, &data);
+#if 0
+                if (verbose > 1) {
+                    data.first = 1;
+                    data.lable = "Key manager";
+                }
+                EVP_KEYMGMT_do_all_ex(NULL, do_keymgmt, &data);
+                if (verbose > 1) {
+                    data.first = 1;
+                    data.lable = "Key exchange";
+                }
+                EVP_KEYEXCH_do_all_ex(NULL, do_keyexch, &data);
+#endif
+
+                switch (verbose) {
+                default:
+                    break;
+                case 2:
+                case 1:
+                    BIO_printf(bio_out, "\n");
+                    break;
+                }
+            }
+            OSSL_PROVIDER_unload(prov);
+        } else {
+            ERR_print_errors(bio_err);
+            /* because exit codes above 127 have special meaning on Unix */
+            if (++ret > 127)
```

Comment:
What the? Is this in any other app?
So if ret was 0 it sets it to 1?
What does this mean when not on unix.

Reply:
This is directly copied from `apps/engine.c`.  So if a provider couldn't be loaded, `ret` gets incremented.  That makes this command usable from a script that wants to see if a set of providers is available.

Windows uses the same type of exit codes, so that's safe.
VMS is a little more special, but since `apps/openssl.c` uses the macro `EXIT`, here's the interesting part for VMS:

https://github.com/openssl/openssl/blob/485d336137f2afa62e378bc39dcfa37dcfb222da/e_os.h#L259-L283

Reply:
It will give the exit code 135, and anyone that does the usual "high bit set means signal" will interpret that as SIGBUS having been thrown (on Linux at least).

But hey, instead of having a bit dispute, I'll just change the code to set `ret = 1`...